### DISCUSSION TITLE
Browsing some directories, all files are hidden

### DISCUSSION THREAD

**Discussion by wookayin** [nvim-neo-tree/neo-tree.nvim#320] [Ideas]
For example, when I open a directory `/opt/homebrew/share/zsh/site-functions/` which is not a git repository, I do not see any files shown in the tree. All the files are treated as hidden files for no reason I can imagine. Is this a bug or expected behavior? How can I tell from which rule neo-tree determines those files are to be hidden?

<img width="877" alt="image" src="https://user-images.githubusercontent.com/1009873/165145033-0dd51d02-9c9b-46b7-b314-b3914c8c9694.png">

<details>
   <summary>My neotree config, just in case (click to expand)</summary>

```
  require('neo-tree').setup {
    filesystem = {
      hijack_netrw_behavior = "open_current",
      window = {
        width = 30,

        mappings = {
          ["r"] = "refresh",
          ["R"] = "rename",
          ["I"] = "toggle_hidden",
          -- Do not reserve default vim keybindings (I want zt, zz, etc.)
          ["z"] = "none",
          ["H"] = "none",
        },
      },

      -- This is a useful feature to turn on,
      -- but should be disabled for a workaround until #111 is fixed
      follow_current_file = false,
      use_libuv_file_watcher = false,

      -- Append trailing slashes on directories (#112)
      components = {
        trailing_slash = function ()
          return {
            text = "/",
            highlight = "NeoTreeDirectoryIcon",
          }
        end,
      },
      renderers = {
        directory = {
          {"icon"},
          {"name", use_git_status_colors = false},
          {"trailing_slash"},
          {"diagnostics"},
          {"git_status"},
        }
      },
    },

    -- Layout (see #130)
    default_component_configs = {
      indent = {
        with_markers = true,
        indent_marker = "│",
        last_indent_marker = "└",
        indent_size = 2,
        padding = 0, -- extra padding on the left hand side
      },
      icon = {
        default = "",
      },
      git_status = {
        symbols = {
          untracked = "?",
          unstaged = "*",
          staged = "✚",
        }
      },
    },
  }
```

</details>

Please feel free to convert this to an issue from a discussion post, if you think it's indeed a bug.

**Comment by cseickel** [comment]
What it is actually doing is running `git check-ignore` within that directory for all the files listed. You can verify whether that ignored status is correct and get an explanation of why they are ignored by running this:

```
git check-ignore -v *
```

Within that directory.

**Reply by wookayin** [reply]
Thanks, I didn't know that useful git command!  It turns out that the directory `/opt/homebrew/` (and its subdirectories, `/opt/homebrew/share/zsh/site-functions/`) is *indeed* a git repository. Yeah, the entire homebrew directories are in a git repository. In the root of the repository there is a `gitignore` file that ignores everything in the subtree:
```
1: # First, ignore everything.
2: /*
```
To my understanding `/*` should match the files in the root only, maybe it applies to `/opt/homebrew/Cellar` and everything under it gets ignored, as `git check-ignore -v*` explains:
```
.gitignore:2:/* (FILES)
```

So the current behavior of neo-tree is kind of works-as-expected, but it's somewhat strange or unexpected from a UX perspective to see no files are shown after navigating to that directory (until you press `I` or toggle showing hidden files). 

<img width="810" alt="image" src="https://user-images.githubusercontent.com/1009873/165166418-4c9c4417-6884-4755-bff7-7bddd9ea1d32.png">

However, it would be reasonable to be able to show hidden files when we have an explicit "root directory" that is also hidden. For instance, `~/.cache` is a folder that should be hidden when listing `~`, but once the root context is narrowed down to `~/.cache` users would expect to see all files under it (which is the behavior of general file explorers, such as Finder in macOS, explorer in windows, to name a few). **Maybe that's when the "root directory" itself is also ignored**, e.g:
```
$ git check-ignore -v .
.gitignore:2:/*	.              # this entire directory is git-ignored !!!
```

It is also partially related to #184. Since the fix the root will never be hidden, but this is about all the subtree being hidden. What do you think would be a reasonable solution or suggested user configs in such cases?

**Reply by cseickel** [reply]
Yes, I think it make sense to not hide gitignored items when you are within a folder that is itself gitignored. That would follow the same logic as the `hide_dotfiles` option.

Another feature to consider, and I'm not sure if this is instead of or along with the above, is that it might be nice to say somewhere that there are hidden items. Maybe just at the bottom of a directory it might say:

(6 hidden items)

**Reply by cseickel** [reply]
Another alternative that is very simple to do and will work in all such situations is to automatically show hidden items in any folder where **all** items would otherwise be filtered out. I would prefer that over the special handling for a git ignored root.

The problem here is that every possible way to handle this is going to be confusing to someone in some specific situation. The trick is trying to anticipate what will cause the least amount of frustration across all applicable situations and thought processes.

No matter what I do about hiding/showing the nodes, I think I need to add the capability of showing a message in unusual situations like this. Either "6 items are hidden" or "All items were hidden, showing them now".

**Comment by cseickel** [comment]
Here's the combo of my last two ideas. I'm showing a message with the count of hidden items in all folders:
![image](https://user-images.githubusercontent.com/5160605/165195198-7ce197d1-6b7a-4d91-9da7-b6d8f7ba3f8b.png)

but then if you navigate into a folder with nothing but hidden items, it forcefully shows them:
![image](https://user-images.githubusercontent.com/5160605/165195258-de144193-0f8b-448d-ad58-a5162c463ecc.png)

What do you think about that? Should it just be one or the other, or is having both features good?

EDIT: For reference, here is that folder when filtered items are visible:
![image](https://user-images.githubusercontent.com/5160605/165195550-67371c60-537d-4ac0-8221-a4a5fbf28ec1.png)

**Reply by nyngwang** [reply]
Wow, this is great!

**Reply by cseickel** [reply]
@nyngwang Do you like both changes or just one of them: That is:

1. would you want to see the number of hidden items in every folder by default?
2. Would you like Neo-tree to automatically show hidden items when you navigate to a root folder that is otherwise empty?

I've already made configuration options for all variations in my local copy, but I still need to decide on what the defaults should be.

**Reply by nyngwang** [reply]
> 1. would you want to see the number of hidden items in every folder by default?

Absolutely yes! One line for such important info. is always a good trade.

> 2. would you like Neo-tree to automatically show hidden items when you navigate to a root folder that is otherwise empty?

I think it's better to always show all the things **by default**. This way no possible information will be out-of-look(the safest option from the developer's perspective). If some people think this is annoying they will search for helps about how to disable this default.

**Reply by wookayin** [reply]
I'm following up the discussions after a while. Wow, this is really great and appreciate all the new features -- with some reasonable defaults :) I like the second option which we now have a configuration point for. Thanks!

**Reply by nyngwang** [reply]
@wookayin: Your good question introduces this good solution!

**Comment by cseickel** [comment]
I quietly release these new features. Without any configuration, it will just show a
```
(# hidden items)
```
message when you browse to a directory with nothing but hidden items in it.

---

If you want to see that hidden items message for all folders that contain hidden items, you can set:

```lua
    require("neo-tree").setup({
      filesystem = {
        filtered_items = {
          show_hidden_count = true, -- when true, the number of hidden items in each folder will be shown as the last entry
        }
      }
    })
```

---

If you want to force showing hidden items in an otherwise empty root folder, you can set:
```lua
    require("neo-tree").setup({
      filesystem = {
        filtered_items = {
          force_visible_in_empty_folder = true, -- when true, hidden files will be shown if the root folder is otherwise empty
        }
      }
    })
```

**Reply by nyngwang** [reply]
I just added it :) This is a very nice feature!
